irods/resource.py

```diff
@@ -38,6 +41,54 @@ def __init__(self, manager, result=None):
 
         self._meta = None
 
+
+    ## Cached properties to expose parent id or name regardless whether the DB model is iRODS 4.1- or 4.2+
+
+    @property
+    def parent_id(self):
+        if self.parent is None:
+            return None
+        if self._parent_id == '':
+            sess = self.manager.sess
+            if sess.server_version >= (4, 2, 0):
+                self._parent_id = self.parent
+            else:
+                self._parent_id = sess.query(Resource).filter(Resource.name == self.parent).one()[Resource.id]
+        return int(self._parent_id)
+
+
+    @property
+    def parent_name(self):
+        if self.parent is None:
+            return None
+        if self._parent_name == '':
+            sess = self.manager.sess
+            if sess.server_version < (4, 2, 0):
+                self._parent_name = self.parent
+            else:
+                self._parent_name = sess.query(Resource).filter(Resource.id == self.parent).one()[Resource.name]
+        return self._parent_name
+
+    ## Cached property to expose resource hierarchy string
+
+    @property
+    def hierarchy(self):
+        if self._hierarchy == '':
+            self._hierarchy = ';'.join(r.name for r in self.node_to_hier())
+        return self._hierarchy
+
+    ## Retrieve chain of parent objects to top level parent
+
+    def node_to_hier(self):
```

Comment:
I have another opinion now that I'm looking at this again...

Is there a reason that this function has to be exposed? Could it just be embedded in `hierarchy`?

Any others have thoughts on this?

Reply:
I think I agree - not too happy with the name (yet).   Keeping it internal seems an easy solution for now.

Reply:
I think embedding it would be fine, yes.  I'll wait on others to weigh in.

Reply:
could we call it _hierarchy_for_node and just imply its hiddenness? this has good precedent.  if you embed it in another function, it causes runtime overhead.

Reply:
...one way in which C++ is definitely Python's superior.

Reply:
So `node_to_hier` returns the hierarchy as a list of `iRODSResource` objects, correct?

Embedding it is a fine solution. Getting the hierarchy as a list of objects seems like a useful thing too (as a convenience).

How do we feel about using more letters/words?
- `hierarchy` -> `hierarchy_string`
- `node_to_hier` -> `hierarchy_as_list_of_resource_objects` (I feel there is a better name hiding in there)

Reply:
"Embedded" perhaps was not the right word. I guess I meant "integrated".

In my opinion, this doesn't need to be a separate method because `hierarchy` is calling it exclusively, and that is the only situation in which this is useful. In fact, I kind of see these as one and the same, which is why I'm suggesting integration.

The primary reason I can see for keeping these separated is for providing multiple options with respect to return types (list versus semicolon-delimited-string). So, should we choose to keep these separated, I think `hierarchy_string` and `hierarchy_as_list_of_resource_objects` definitely conveys what this method is doing very clearly. 

Alternatively, `hierarchy` could be a class (or sub-class) which provides a `__str__()` method to get the string version. Or something.

Reply:
Ok,  I renamed the method, which I think is as much as the present issue demands.  If there's a perceived need for hierarchy to be its own concept (with class and stringization methods), then I can make a new issue.  But I think this fills the bill for now.

Reply:
A last note - I took the present approach - with "B being called exclusively by A" - as being a way to simplify the logical flow of A and make it more readable.

Reply:
I think the implementation in the latest commit is fine. Let's hold off on the `hierarchy` type idea since this implementation working and clear.